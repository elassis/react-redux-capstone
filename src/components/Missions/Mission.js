import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { setReserved, setUnReserved } from '../../redux/missions/missions';
import classes from './Missions.module.css';

const Mission = ({ mission }) => {
  const {
    id, name, description, reserved,
  } = mission;
  const [buttonContent, setButtonContent] = useState('Join Mission');
  const [status, setStatus] = useState('NOT A MEMBER');
  const dispatch = useDispatch();

  const handleJoinMission = () => {
    if (!reserved) {
      setButtonContent('Leave Mission');
      setStatus('Active Member');
      dispatch(setReserved(id));
    } else if (reserved) {
      setButtonContent('Join Mission');
      setStatus('NOT A MEMBER');
      dispatch(setUnReserved(id));
    }
  };

  return (
    <tr className={classes.tableRow}>
      <td className={classes.tableData}>{name}</td>
      <td className={classes.tableData}>{description}</td>
      <td className={classes.tableData}>
        <span className={`${classes.tableDataStatus} ${!reserved ? classes.statusActive : classes.statusInactive}`}>
          {status}
        </span>
      </td>
      <td className={`${classes.tableData} ${classes.tableDataBtn}`}>
        <button className={!reserved ? classes.joinMissionBtn : classes.leaveMissionBtn} onClick={handleJoinMission} type="button">
          {buttonContent}
        </button>
      </td>
    </tr>
  );
};

Mission.defaultProps = {
  mission: {
    id: '',
    name: '',
    description: '',
    reserved: '',
  },
};

Mission.propTypes = {
  mission: PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    reserved: PropTypes.bool,
  }),
};

export default Mission;
