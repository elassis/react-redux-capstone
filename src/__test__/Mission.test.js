import React from 'react';
import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../redux/configureStore';
import Mission from '../components/missions/Mission';
import '@testing-library/jest-dom/extend-expect';

describe('Missions', () => {
  test('renders component correctly', () => {
    const tree = render(
      <Provider store={store}>
        <Router>
          <table>
            <thead />
            <tbody>
              <Mission />
            </tbody>
          </table>
        </Router>
      </Provider>,
    );
    expect(tree).toMatchSnapshot();
  });

  test('renders Missions component', () => {
    render(
      <Provider store={store}>
        <Router>
          <table>
            <thead />
            <tbody>
              <Mission />
            </tbody>
          </table>
        </Router>
      </Provider>,
    );

    expect(screen.getByText(/Join Mission/)).toBeInTheDocument();
  });
});
