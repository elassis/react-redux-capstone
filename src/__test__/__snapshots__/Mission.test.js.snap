// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions renders component correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table>
        <thead />
        <tbody>
          <tr
            class=""
          >
            <td
              class="tableData"
            />
            <td
              class="tableData"
            />
            <td
              class="tableData"
            >
              <span
                class="tableDataStatus statusActive"
              >
                NOT A MEMBER
              </span>
            </td>
            <td
              class="tableData tableDataBtn"
            >
              <button
                class="joinMissionBtn"
                type="button"
              >
                Join Mission
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table>
      <thead />
      <tbody>
        <tr
          class=""
        >
          <td
            class="tableData"
          />
          <td
            class="tableData"
          />
          <td
            class="tableData"
          >
            <span
              class="tableDataStatus statusActive"
            >
              NOT A MEMBER
            </span>
          </td>
          <td
            class="tableData tableDataBtn"
          >
            <button
              class="joinMissionBtn"
              type="button"
            >
              Join Mission
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
